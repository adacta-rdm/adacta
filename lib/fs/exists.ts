import { access } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export async function exists(path: PathArg) {
	try {
		await access(normalizePath(path));
		return true;
	} catch (e) {
		return false;
	}
}
