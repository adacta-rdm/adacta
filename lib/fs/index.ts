/**
 * @file Automatically generated by barrelsby.
 */

export * from "./exists";
export * from "./isDir";
export * from "./mkdirTmp";
export * from "./mkdirp";
export * from "./readFile";
export * from "./readJSONFile";
export * from "./readUTF8File";
export * from "./readdir";
export * from "./rmrf";
export * from "./stat";
export * from "./writeFile";
export * from "./private/normalizePath";
