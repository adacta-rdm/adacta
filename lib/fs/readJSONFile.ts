import type { PathArg } from "./private/normalizePath";
import { readUTF8File } from "./readUTF8File";

export function readJSONFile(path: PathArg): Promise<unknown> {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
	return readUTF8File(path).then((fileContents) => JSON.parse(fileContents));
}
