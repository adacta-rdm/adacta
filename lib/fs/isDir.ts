import { stat } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export async function isDir(path: PathArg) {
	try {
		const stats = await stat(normalizePath(path));
		return stats.isDirectory();
	} catch (e: any) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
		if (e.code && e.code === "ENOENT") {
			return false;
		}
		throw e;
	}
}
