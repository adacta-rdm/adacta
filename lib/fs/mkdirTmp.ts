import { mkdtemp } from "fs/promises";
import { tmpdir } from "os";
import { join } from "path";

export function mkdirTmp(prefix = "") {
	return mkdtemp(join(tmpdir(), "/", prefix));
}
