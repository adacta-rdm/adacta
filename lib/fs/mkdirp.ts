import { mkdir } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export async function mkdirp(path: PathArg) {
	return mkdir(normalizePath(path), { recursive: true });
}
