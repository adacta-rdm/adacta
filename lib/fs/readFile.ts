import { readFile as readFileFs } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export function readFile(path: PathArg): Promise<Buffer> {
	return readFileFs(normalizePath(path));
}
