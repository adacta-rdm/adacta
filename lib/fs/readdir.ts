import { readdir as readdirFs } from "fs/promises";

import type { PathArg } from "./private/normalizePath";
import { normalizePath } from "./private/normalizePath";

export function readdir(path: PathArg) {
	return readdirFs(normalizePath(path));
}
