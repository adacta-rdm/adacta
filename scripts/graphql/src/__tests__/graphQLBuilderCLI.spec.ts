import { readFileSync } from "node:fs";
import { dirname, join } from "path";

import { mkdirp, mkdirTmp, readdir, readUTF8File, rmrf, writeFile } from "@omegadot/fs";
import { describe, test, expect, afterEach } from "vitest";
import { vi as jest } from "vitest";

import { graphQLBuilderCLI } from "../graphQLBuilderCLI";

const MyType2 = readFileSync(join(__dirname, "./typeDefs/MyType2.graphql"), "utf-8");
const queries = readFileSync(join(__dirname, "./typeDefs/queries.graphql"), "utf-8");
const schema = readFileSync(join(__dirname, "./typeDefs/schema.graphql"), "utf-8");
const schema1 = readFileSync(join(__dirname, "./typeDefs/schema1.graphql"), "utf-8");
const types = readFileSync(join(__dirname, "./typeDefs/types.graphql"), "utf-8");

describe("graphQLBuilderCLI()", () => {
	async function setup() {
		const tmpDir = await mkdirTmp();

		await mkdirp([tmpDir, "graphql"]);

		await writeFile([tmpDir, "graphql/schema1.graphql"], schema1);
		await writeFile([tmpDir, "graphql/MyType2.graphql"], MyType2);
		await writeFile(
			//
			[tmpDir, ".graphqlconfig"],
			JSON.stringify({
				projects: {
					Test1: {
						includes: ["graphql/*.graphql"],
					},
				},
			})
		);

		return {
			tmpDir,
			include: join(tmpDir, "graphql/*.graphql"),
			outdir: join(tmpDir, "generated"),
		};
	}

	afterEach(() => {
		jest.restoreAllMocks();
	});

	test("single file build", async () => {
		const { include, tmpDir } = await setup();

		const outfile = join(tmpDir, "schema.graphql");

		await graphQLBuilderCLI({ include, build: outfile });

		const generated = await readUTF8File(outfile);

		expect(generated).toContain("type Result_Type1");
		expect(generated).toContain("type Result_Type2");
		expect(generated).toContain("type Result_T");
		expect(generated).toContain("scalar T");
	});

	test("single file build - project option", async () => {
		const { tmpDir } = await setup();

		jest.spyOn(process, "cwd").mockImplementation(() => tmpDir);

		const outfile = join(tmpDir, "schema.graphql");

		await graphQLBuilderCLI({ project: "Test1", build: outfile });

		const generated = await readUTF8File(outfile);

		expect(generated).toContain("type Result_Type1");
		expect(generated).toContain("type Result_Type2");
		expect(generated).toContain("type Result_T");
		expect(generated).toContain("scalar T");
	});

	test("bundle", async () => {
		// Similar to a single file build, but the output file includes only the types that are actually used by the
		// entrypoint type.
		const { include, tmpDir } = await setup();
		await writeFile([tmpDir, "graphql/schema.graphql"], schema);
		const outfile = join(tmpDir, "schema-bundle.graphql");

		await graphQLBuilderCLI({ include, bundle: outfile });

		const generated = await readUTF8File(outfile);

		expect(generated).toContain("schema");
		expect(generated).toContain("type Query");
		expect(generated).toContain("type MyType");
		expect(generated).toContain("type Result_Type1");
		expect(generated).not.toContain("interface Result_T");
	});

	test("multiple file build", async () => {
		const { include, outdir } = await setup();

		await graphQLBuilderCLI({ include, buildDir: outdir });

		const list = await readdir(outdir);

		expect(list).toEqual(["Result_Type1.graphql", "Result_Type2.graphql", "T.graphql"]);
	});

	test("build - deletes generated files that are obsolete", async () => {
		// Generate the files once, then delete the MyType2.graphql file. Because that type is the only one that uses
		// Result_Type2, the file Result_Type2.graphql should be deleted as well.
		const { tmpDir, include, outdir } = await setup();

		await graphQLBuilderCLI({ include, buildDir: outdir });

		expect(await readdir(outdir)).toContain("Result_Type2.graphql");

		await rmrf([tmpDir, "graphql/MyType2.graphql"]);

		await graphQLBuilderCLI({ include, buildDir: outdir });
		expect(await readdir(outdir)).not.toContain("Result_Type2.graphql");
	});

	test("build - deletes generated files that are obsolete when writing to include dir", async () => {
		// Generate the files once, then delete the MyType2.graphql file. Because that type is the only one that uses
		// Result_Type2, the file Result_Type2.graphql should be deleted as well.
		const { tmpDir, include } = await setup();

		const outdir = dirname(include);

		await graphQLBuilderCLI({ include, buildDir: outdir });

		expect(await readdir(outdir)).toContain("Result_Type2.graphql");

		await rmrf([tmpDir, "graphql/MyType2.graphql"]);

		await graphQLBuilderCLI({ include, buildDir: outdir });
		expect(await readdir(outdir)).not.toContain("Result_Type2.graphql");
	});

	test("multiple file build - overwrites only generated files", async () => {
		// Generate the files once, then modify two of them, keeping the @generated comment only in one of them.
		// Run the script again and verify that only the one with the comment is overwritten.
		const { include, outdir } = await setup();

		await graphQLBuilderCLI({ include, buildDir: outdir });

		const fileContents1 = `type Result_Type1 { id: ID! }\n# This file was modified.`;
		const fileContents2 = `# @generated\n\ntype Result_Type2 { id: ID! }\n# This file was modified.`;
		await writeFile([outdir, "Result_Type1.graphql"], fileContents1);
		await writeFile([outdir, "Result_Type2.graphql"], fileContents2);

		await graphQLBuilderCLI({ include, buildDir: outdir });

		expect(await readUTF8File([outdir, "Result_Type1.graphql"])).toBe(fileContents1);
		// expect(await readUTF8File([outdir, "Result_Type2.graphql"])).not.toBe(fileContents2);
	});

	test("typescript type generation - resolver types", async () => {
		const { tmpDir, include, outdir } = await setup();
		// This file contains a schema definition needed for the bundle option.
		await writeFile([tmpDir, "graphql/schema.graphql"], schema);

		const outFileTsResolvers = join(outdir, "types.ts");

		await graphQLBuilderCLI({ include, tsSchema: outFileTsResolvers });

		const list = await readdir(outdir);

		expect(list).toEqual(["types.ts"]);
		expect(await readUTF8File(outFileTsResolvers)).toMatchSnapshot();
	});

	test("typescript type generation - requests code", async () => {
		const tmpDir = await mkdirTmp();
		// This file contains a schema definition needed for the bundle option.
		// await writeFile([tmpDir, "graphql/schema.graphql"], schema);
		await writeFile([tmpDir, "queries.graphql"], queries);
		await writeFile([tmpDir, "types.graphql"], types);

		const outFileTsRequests = join(tmpDir, "requests.ts");

		await graphQLBuilderCLI({ include: join(tmpDir, "*.graphql"), tsRequests: outFileTsRequests });

		const list = await readdir(tmpDir);

		expect(list).toContain("requests.ts");
		expect(await readUTF8File(outFileTsRequests)).toMatchSnapshot();
	});
});
