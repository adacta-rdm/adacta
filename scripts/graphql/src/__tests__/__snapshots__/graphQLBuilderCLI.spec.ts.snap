// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphQLBuilderCLI() > typescript type generation - requests code 1`] = `
"// @generated
import type { GraphQLClient, RequestOptions } from "graphql-request";
import gql from "graphql-tag";
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
  [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
  [SubKey in K]: Maybe<T[SubKey]>;
};
type GraphQLClientRequestHeaders = RequestOptions["requestHeaders"];
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type IQuery = {
  __typename?: "Query";
  books: IConnection_Book;
  book?: Maybe<IBook>;
  func?: Maybe<Scalars["String"]>;
};

export type IQueryBookArgs = {
  id: Scalars["ID"];
};

export type IQueryFuncArgs = {
  input?: InputMaybe<IPartial_Input1>;
};

export type IConnection_Book = {
  __typename?: "Connection_Book";
  edges: Array<IEdge_Book>;
};

export type IEdge_Book = {
  __typename?: "Edge_Book";
  node: IBook;
};

export type IBook = {
  __typename?: "Book";
  id: Scalars["ID"];
  wrappedString: IWrapper_String;
};

export type IWrapper_String = {
  __typename?: "Wrapper_String";
  value: Scalars["String"];
};

export type IPartial_Input1 = {
  prop1?: InputMaybe<Scalars["Float"]>;
  prop2?: InputMaybe<Scalars["String"]>;
  prop3?: InputMaybe<Scalars["Int"]>;
};

export type IQuery1QueryVariables = Exact<{
  input: Scalars["ID"];
}>;

export type IQuery1Query = {
  __typename?: "Query";
  book?: {
    __typename?: "Book";
    id: string;
    wrappedString: { __typename?: "Wrapper_String"; value: string };
  } | null;
};

export const Query1Document = gql\`
  query Query1($input: ID!) {
    book(id: $input) {
      id
      wrappedString {
        value
      }
    }
  }
\`;

export type SdkFunctionWrapper = <T>(
  action: (requestHeaders?: Record<string, string>) => Promise<T>,
  operationName: string,
  operationType?: string,
  variables?: any
) => Promise<T>;

const defaultWrapper: SdkFunctionWrapper = (
  action,
  _operationName,
  _operationType,
  _variables
) => action();

export function getSdk(
  client: GraphQLClient,
  withWrapper: SdkFunctionWrapper = defaultWrapper
) {
  return {
    Query1(
      variables: IQuery1QueryVariables,
      requestHeaders?: GraphQLClientRequestHeaders
    ): Promise<IQuery1Query> {
      return withWrapper(
        (wrappedRequestHeaders) =>
          client.request<IQuery1Query>(Query1Document, variables, {
            ...requestHeaders,
            ...wrappedRequestHeaders,
          }),
        "Query1",
        "query",
        variables
      );
    },
  };
}
export type Sdk = ReturnType<typeof getSdk>;
"
`;

exports[`graphQLBuilderCLI() > typescript type generation - resolver types 1`] = `
"// @generated
import { ResolverReturnType } from "~/lib/utils/types";
import type { GraphQLResolveInfo } from "graphql";
import type { IGraphQLContext } from "../IGraphQLContext";
export type Maybe<T> = T | null | undefined;
export type InputMaybe<T> = T | null | undefined;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
  [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
  [SubKey in K]: Maybe<T[SubKey]>;
};
export type RequireFields<T, K extends keyof T> = Omit<T, K> & {
  [P in K]-?: NonNullable<T[P]>;
};
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type IQuery = {
  __typename?: "Query";
  /** Returns "world". */
  hello: Scalars["String"];
  myType?: Maybe<IMyType>;
};

export type IQueryMyTypeArgs = {
  id: Scalars["ID"];
};

export type IMyType = {
  __typename?: "MyType";
  result?: Maybe<IResult_Type1>;
};

export type IResult_Type1 = INode & {
  __typename?: "Result_Type1";
  id: Scalars["ID"];
  payload?: Maybe<IType1>;
};

export type INode = {
  id: Scalars["ID"];
};

export type IType1 = {
  __typename?: "Type1";
  name?: Maybe<Scalars["String"]>;
};

export type ResolverTypeWrapper<T> = Promise<T> | T;

export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> =
  | ResolverFn<TResult, TParent, TContext, TArgs>
  | ResolverWithResolve<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<
  TResult,
  TKey extends string,
  TParent,
  TContext,
  TArgs
> {
  subscribe: SubscriptionSubscribeFn<
    { [key in TKey]: TResult },
    TParent,
    TContext,
    TArgs
  >;
  resolve?: SubscriptionResolveFn<
    TResult,
    { [key in TKey]: TResult },
    TContext,
    TArgs
  >;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<
  TResult,
  TKey extends string,
  TParent,
  TContext,
  TArgs
> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<
  TResult,
  TKey extends string,
  TParent = {},
  TContext = {},
  TArgs = {}
> =
  | ((
      ...args: any[]
    ) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (
  obj: T,
  context: TContext,
  info: GraphQLResolveInfo
) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<
  TResult = {},
  TParent = {},
  TContext = {},
  TArgs = {}
> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

/** Mapping between all available schema types and the resolvers types */
export type IResolversTypes = {
  Query: ResolverTypeWrapper<{}>;
  String: ResolverTypeWrapper<ResolverReturnType<Scalars["String"]>>;
  ID: ResolverTypeWrapper<ResolverReturnType<Scalars["ID"]>>;
  MyType: ResolverTypeWrapper<ResolverReturnType<IMyType>>;
  Result_Type1: ResolverTypeWrapper<ResolverReturnType<IResult_Type1>>;
  Node: IResolversTypes["Result_Type1"];
  Type1: ResolverTypeWrapper<ResolverReturnType<IType1>>;
  Boolean: ResolverTypeWrapper<ResolverReturnType<Scalars["Boolean"]>>;
};

/** Mapping between all available schema types and the resolvers parents */
export type IResolversParentTypes = {
  Query: {};
  String: ResolverReturnType<Scalars["String"]>;
  ID: ResolverReturnType<Scalars["ID"]>;
  MyType: ResolverReturnType<IMyType>;
  Result_Type1: ResolverReturnType<IResult_Type1>;
  Node: IResolversParentTypes["Result_Type1"];
  Type1: ResolverReturnType<IType1>;
  Boolean: ResolverReturnType<Scalars["Boolean"]>;
};

export type IQueryResolvers<
  ContextType = IGraphQLContext,
  ParentType extends IResolversParentTypes["Query"] = IResolversParentTypes["Query"]
> = {
  hello?: Resolver<IResolversTypes["String"], ParentType, ContextType>;
  myType?: Resolver<
    Maybe<IResolversTypes["MyType"]>,
    ParentType,
    ContextType,
    RequireFields<IQueryMyTypeArgs, "id">
  >;
};

export type IMyTypeResolvers<
  ContextType = IGraphQLContext,
  ParentType extends IResolversParentTypes["MyType"] = IResolversParentTypes["MyType"]
> = {
  result?: Resolver<
    Maybe<IResolversTypes["Result_Type1"]>,
    ParentType,
    ContextType
  >;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type IResult_Type1Resolvers<
  ContextType = IGraphQLContext,
  ParentType extends IResolversParentTypes["Result_Type1"] = IResolversParentTypes["Result_Type1"]
> = {
  id?: Resolver<IResolversTypes["ID"], ParentType, ContextType>;
  payload?: Resolver<Maybe<IResolversTypes["Type1"]>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type INodeResolvers<
  ContextType = IGraphQLContext,
  ParentType extends IResolversParentTypes["Node"] = IResolversParentTypes["Node"]
> = {
  __resolveType: TypeResolveFn<"Result_Type1", ParentType, ContextType>;
  id?: Resolver<IResolversTypes["ID"], ParentType, ContextType>;
};

export type IType1Resolvers<
  ContextType = IGraphQLContext,
  ParentType extends IResolversParentTypes["Type1"] = IResolversParentTypes["Type1"]
> = {
  name?: Resolver<Maybe<IResolversTypes["String"]>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type IResolvers<ContextType = IGraphQLContext> = {
  Query?: IQueryResolvers<ContextType>;
  MyType?: IMyTypeResolvers<ContextType>;
  Result_Type1?: IResult_Type1Resolvers<ContextType>;
  Node?: INodeResolvers<ContextType>;
  Type1?: IType1Resolvers<ContextType>;
};
"
`;
